### PR TITLE
Engineering GUI GSAS tab add Refine All button

### LINKED ISSUE #22265: Engg Diffraction GUI - GSAS tab should have a 'fit all' button
The GSAS tab should have a **Fit all** button which runs refinement on every run loaded into the GUI, in the same way as the Fitting tab

### PR DESCRIPTION
**Pull request by josephframsay**
A button to run refinement on all runs loaded into the GSAS tab of the ED GUI has been added.

**To test:**

Some sample data is [here](https://github.com/mantidproject/mantid/files/1894002/GSASIIRefineFitPeaks.zip). You'll also need GSASII installed - go [here](http://docs.mantidproject.org/nightly/algorithms/GSASIIRefineFitPeaks-v1.html#installing-gsasii) for how to install GSASII.

Select the two Nexus files for **Focused Run Files** and click **Load**. Select the `.prm` file for **Instrument Parameter File**, the three `.cif`s for **Phase Files**, the location of `GSASIIscriptable.py` (part of GSASII) for **GSAS-II Installation Directory**, and choose somewhere to save out the GSAS-II project files for **New GSAS-II Project**. Fiddle with any other parameters (keep **RefinementMethod** as Rietveld), and click **Refine All**.

Make sure a refinement was completed for every run (you should see values in the **Fit Results** section and overplotted fit results in the **Preview** section. Try cancelling the algorithm and closing the GUI while refinement is running.

Fixes #22265 

**Release Notes** 
[Here](https://github.com/mantidproject/mantid/commit/cd6a9ac0278e2e00ab5b389d4d421544880bc4c8)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [x] Are the unit tests small and test the class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] Do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/EnggDiffraction/EnggDiffGSASFittingPresenter.cpp [hunk 1]
```diff
@@ -45,6 +63,47 @@ void EnggDiffGSASFittingPresenter::notify(
...
+  const auto refinementMethod = m_view->getRefinementMethod();
+  const auto instParamFile = m_view->getInstrumentFileName();
+  const auto phaseFiles = m_view->getPhaseFileNames();
+  const auto pathToGSASII = m_view->getPathToGSASII();
+  const auto GSASIIProjectFile = m_view->getGSASIIProjectPath();
+  if (runLabels.size() == 1) {
+    GSASIIProjectFiles = std::vector<std::string>({GSASIIProjectFile});
+  } else {
+    for (const auto &runLabel : runLabels) {
+      GSASIIProjectFiles.push_back(
```
Comment (soininen) [inline qt/scientific_interfaces/EnggDiffraction/EnggDiffGSASFittingPresenter.cpp]: `emplace_back` should work at least as efficiently as `push_back`.

File: qt/scientific_interfaces/EnggDiffraction/EnggDiffMultiRunFittingQtWidget.cpp [hunk 2]
```diff
@@ -33,17 +49,22 @@ void EnggDiffMultiRunFittingQtWidget::cleanUpPlot() {
   m_fittedPeaksCurves.clear();
 }
 
+std::vector<RunLabel> EnggDiffMultiRunFittingQtWidget::getAllRunLabels() const {
+  std::vector<RunLabel> runLabels;
+  runLabels.reserve(m_ui.listWidget_runLabels->count());
+
+  for (int i = 0; i < m_ui.listWidget_runLabels->count(); i++) {
+    const auto currentLabel = m_ui.listWidget_runLabels->item(i)->text();
+    runLabels.push_back(parseListWidgetItem(currentLabel));
```
Comment (soininen) [inline qt/scientific_interfaces/EnggDiffraction/EnggDiffMultiRunFittingQtWidget.cpp]: `emplace_back`

**Review by soininen — CHANGES_REQUESTED** [review]
Looking good, some unit tests are added. Just a few minor comments below.

**Comment by soininen** [comment]
The double free corruption when closing the interface/MantidPlot is now gone. Tested the interface with the provided dataset and everything seems to work now as expected. Good job!

**Review by soininen — APPROVED** [review]

**Comment by martyngigg** [comment]
The OSX test failure is unrelated.

**Pull request merged**